Premium Blocks: tracking events

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

This PR records two new tracking events: `jetpack_editor_block_upgrade_click` and `jetpack_editor_block_upgrade_banner_impression`.

Fixes #16863 

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Premium Blocks: record tracking events.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

Yes.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply D48278-code
* Get sandboxed
* Pick up a Simple site to test (Sandbox it!)
<img src="https://user-images.githubusercontent.com/77539/90570798-d98d3a00-e186-11ea-94f4-990ce840258f.png" width="400px" />


* set debug namespace:
`localStorage.setItem( 'debug', 'dops:analytics*' );`
* add a premium block, for instance `calendly`

<img src="https://user-images.githubusercontent.com/77539/90570848-f6297200-e186-11ea-9c25-540e29da0d73.png" width="600px">

* You should see the first event recorded `jetpack_editor_block_upgrade_banner_impression`
![image](https://user-images.githubusercontent.com/77539/90570906-10635000-e187-11ea-9331-d15c009873b6.png)

* Confirm you see the `context`, `block` and `plan`

* Click on the Upgrade button. Activate preserve logs in order to don't lose it.

<img width="893" alt="Screen Shot 2020-08-18 at 7 17 43 PM" src="https://user-images.githubusercontent.com/77539/90571133-8ff11f00-e187-11ea-862c-7769d058b16f.png">

![image](https://user-images.githubusercontent.com/77539/90571156-9a131d80-e187-11ea-9ee1-2fbd6433bc5d.png)

* Click on the sidebar banner, confirming that now the context is `sidebar`.

![image](https://user-images.githubusercontent.com/77539/90571245-c038bd80-e187-11ea-998d-a37ecbe88796.png)


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Premium Blocks: record tracking events.

retrofox committed:
paid-blocks: add utils tracking helper fn

retrofox committed:
upgrade-banner: track button click event

retrofox committed:
upgrade-banner: record banner impresion event

retrofox committed:
paid-bolocks: reformat code

cpapazoglou commented on extensions/extended-blocks/paid-blocks/with-upgrade-banner.jsx:
```diff
@@ -33,6 +34,22 @@ export default createHigherOrderComponent(
+			context: bannerContext,
+		};
+
+		useEffect( () => {
+			if ( ! isVisible ) {
+				return;
+			}
+
+			trackUpgradeBannerImpression( trackEventData );
+		}, [ isVisible, trackEventData ] );
```
It seems that the event `jetpack_editor_block_upgrade_banner_impression` is fired also when clicking outside the block. Also it is fired on page reload without the banner being shown. Using `isBannerVisible` as dependency instead of `isVisible` seems to resolve that issue for me.

As a note also, It seems that the event `jetpack_editor_block_upgrade_banner_impression` is fired every-time the banner is visible. So clicking outside the block and then inside triggers the event. I believe we should only track the event once? Can't see the value of tracking the banner impression many times in such a short period of time but maybe I am missing something. cc @davemart-in who maybe can enlighten us.

![SS 2020-08-19 at 15 10 11](https://user-images.githubusercontent.com/12430020/90633181-43acea00-e22e-11ea-9088-17408119888c.gif)

retrofox replied:
Nice 👀 , Harris as usual 👏 

`IsBannerVisible` has been addressed [here](https://github.com/Automattic/jetpack/pull/16882/commits/70fa5e00c2f0addc323a1f84715f46cbe370f2b8).

Asking to @lessbloat: should we record the `jetpack_editor_block_upgrade_banner_impression` track-event every time that the banner shows up, or just once, in the first time.
The old Upgrade Nudge was visible constantly so it used to record the event just once.

davemart-in replied:
Just once. Thanks!

retrofox replied:
Addressed [here](33a9bde9cb974111fe3f39164612cda15bd5f407)

cpapazoglou reviewed (changes requested):
Tested by syncing the code into my sandbox (the diff patch was outdated) and can confirm that tracking events work as advertised

![](https://cln.sh/J3E7Bv+)

Also, confirmed that tracking events are not fired once I upgraded my plan.

Just left a comment for multiple `jetpack_editor_block_upgrade_banner_impression` events that seem unwanted and we need to address.

As a note, which seems correct in my opinion:
`jetpack_editor_block_upgrade_banner_impression` runs only for the `editor-canvas` 
due to the fact that both `editor-canvas` and `sidebar` appear simultaneously I believe that this is correct so we avoid duplicate impressions.

retrofox committed:
upgrade-banner: record event only when is vissible
Props to @cpapazoglou to report it https://github.com/Automattic/jetpack/pull/16882#pullrequestreview-470394178

retrofox committed:
upgrade-banner: record impression event once

retrofox committed:
upgrade-banner: fix flag issue

retrofox committed:
upgrade-banner: fix recording condition again \m/

cpapazoglou reviewed (approved):
Passing this once more after the changes, I can confirm that:
- the tracking event is fired only if the banner is visible
- the tracking event is fired only once per banner being visible to the user

cpapazoglou commented:
Passing this issue to jetpack crew for review!

zinigor commented:
Thanks for adding these events! The events work fine, but I have found a couple of problems with the banner itself, is this a good place to fix them?

retrofox commented:
> Thanks for adding these events!

Thanks, Igor. 👍 

> ... but I have found a couple of problems with the banner itself, is this a good place to fix them?

 I suggest creating a new issue and adding the `[Extensions] Paid Blocks` label on.

zinigor reviewed (approved):
The PR works fine, thank you! I have created #16912 to track the other problems.

retrofox commented:
r212431-wpcom
